libraries/Microsoft.Bot.Builder.Core.State/FileSystemStateStorageProvider.cs

```diff
@@ -0,0 +1,330 @@
+        public async Task<IEnumerable<IStateStorageEntry>> Load(string stateNamespace, IEnumerable<string> keys)
+        {
+            var loadTasks = new List<Task<IStateStorageEntry>>();
+
+            foreach (var key in keys)
+            {
+                loadTasks.Add(Load(stateNamespace, key));
+            }
+
+            await Task.WhenAll(loadTasks);
+
+            var results = new List<IStateStorageEntry>(loadTasks.Count);
+
+            foreach (var loadTask in loadTasks)
+            {
+                results.Add(loadTask.Result);
+            }
+
+            return results;
+        }
+
+        public Task Save(IEnumerable<IStateStorageEntry> stateStorageEntries)
+        {
+            var saveTasks = new List<Task>();
+
+            foreach (var stateStorageEntry in stateStorageEntries)
+            {
+                if (!(stateStorageEntry is FileSystemStateStorageEntry fileSystemStateStorageEntry))
+                {
+                    throw new InvalidOperationException($"Only instances of {nameof(FileSystemStateStorageEntry)} are supported by {nameof(FileSystemStateStorageProvider)}.");
+                }
+
+                saveTasks.Add(fileSystemStateStorageEntry.WriteToFile());
+            }
+
+            return Task.WhenAll(saveTasks);
+        }
+
+        private DirectoryInfo GetStateNamespaceDirectoryInfo(string stateNamespace) => new DirectoryInfo(ConvertStateNamespaceToDirectoryPath(_basePath, stateNamespace));
+
+        private static FileInfo GetStateStorageEntryFileInfo(string basePath, string stateNamespace, string key) => new FileInfo(GetStateStorageEntryFilePath(basePath, stateNamespace, key));
+
+        private static string GetStateStorageEntryFilePath(string basePath, string stateNamespace, string key) => Path.Combine(ConvertStateNamespaceToDirectoryPath(basePath, stateNamespace), ConvertKeyToFileName(key));
+
+        public static string ConvertStateNamespaceToDirectoryPath(string basePath, string stateNamespace)
+        {
+            if (string.IsNullOrEmpty(stateNamespace))
+            {
+                throw new ArgumentException("Expected a non-null/empty value.", nameof(stateNamespace));
+            }
+            // Strip any leading path separator so that it always becomes relative to basePath
+            if (stateNamespace[0] == Path.DirectorySeparatorChar || stateNamespace[0] == Path.AltDirectorySeparatorChar)
+            {
+                stateNamespace = stateNamespace.Substring(1);
+            }
+
+            foreach (var invalidPathChar in Path.GetInvalidPathChars())
+            {
+                stateNamespace = stateNamespace.Replace(invalidPathChar, '`');
+            }
+
+            return Path.Combine(basePath, stateNamespace);
+        }
+
+        public static string ConvertKeyToFileName(string key)
+        {
+            foreach (var invalidPathChar in Path.GetInvalidFileNameChars())
+            {
+                key = key.Replace(invalidPathChar, '@');
```

Comment:
That's a totally valid point about collisions I would say that disallowing them in the provider itself would lead to provider specific nuances that would violate the abstraction. 

Maybe the answer is we just black list a bunch of these characters at the state manager key level? I'm totally open to better suggestions. I would suggest that it shouldn't hold up the PR. We've already agree (comment above) we _NEED_ to put in an issue in to come back and harden this entire implementation later, this could just be one of the bullet points on that issue. You're call, LMK.